e2e/image/push_trust_test.go

```diff
@@ -0,0 +1,160 @@
+	defer dir.Remove()
+	image := createImage(t, registryPrefix, "trust-push-unreachable", "latest")
+
+	result := icmd.RunCmd(icmd.Command("docker", "push", image),
+		fixtures.WithConfig(dir.Path()),
+		fixtures.WithTrust,
+		fixtures.WithNotaryServer("https://invalidnotaryserver"),
+	)
+	result.Assert(t, icmd.Expected{
+		ExitCode: 1,
+		Err:      "error contacting notary server",
+	})
+
+	// FIXME(vdemeester) doesn't work
+	/*
+		// Should also fail if `--disable-content-trust` is set
+		result = icmd.RunCmd(icmd.Command("docker", "push", "--disable-content-trust", image),
+			fixtures.WithConfig(dir.Path()),
+			fixtures.WithTrust,
+			fixtures.WithNotaryServer("https://invalidnotaryserver"),
+		)
+		result.Assert(t, icmd.Expected{
+			ExitCode: 1,
+			Err:      "Error establishing connection to notary repository",
+		})
+	*/
+}
+
+func TestPushWithContentTrustExistingTag(t *testing.T) {
+	dir := fixtures.SetupConfigFile(t)
+	defer dir.Remove()
+	image := createImage(t, registryPrefix, "trust-push-existing", "latest")
+
+	result := icmd.RunCmd(icmd.Command("docker", "push", image))
+	result.Assert(t, icmd.Success)
+
+	result = icmd.RunCmd(icmd.Command("docker", "push", image),
+		fixtures.WithConfig(dir.Path()),
+		fixtures.WithTrust,
+		fixtures.WithNotary,
+		fixtures.WithPassphrase("foo", "bar"),
+	)
+	result.Assert(t, icmd.Expected{
+		Out: "Signing and pushing trust metadata",
+	})
+
+	// Re-push
+	result = icmd.RunCmd(icmd.Command("docker", "push", image),
+		fixtures.WithConfig(dir.Path()),
+		fixtures.WithTrust,
+		fixtures.WithNotary,
+		fixtures.WithPassphrase("foo", "bar"),
+	)
+	result.Assert(t, icmd.Expected{
+		Out: "Signing and pushing trust metadata",
+	})
+
+	// FIXME(vdemeester) doesn't work
+	/*
+		// With a wrong password
+		result = icmd.RunCmd(icmd.Command("docker", "push", image),
+			fixtures.WithConfig(dir.Path()),
+			fixtures.WithTrust,
+			fixtures.WithNotary,
+			fixtures.WithPassphrase("bar", "baz"),
+		)
+		result.Assert(t, icmd.Expected{
+			ExitCode: 1,
+			Err:      "could not find necessary signing keys",
+		})
+	*/
+}
+
+func TestPushWithContentTrustReleasesDelegationOnly(t *testing.T) {
+	dir := fixtures.SetupConfigFile(t)
+	defer dir.Remove()
+	notaryDir := setupNotaryConfig(t, dir)
+	defer notaryDir.Remove()
+
+	image := createImage(t, registryPrefix, "trust-push-releases-delegation", "latest")
+
+	icmd.RunCmd(
+		icmd.Command(notary, "-c", notaryDir.Path(), "init", image),
+		withNotaryPassphrase("foo"),
+	).Assert(t, icmd.Success)
+
+	pubkey := "./testdata/notary/delgkey1.crt"
+	icmd.RunCmd(
+		icmd.Command(notary, "-c", notaryDir.Path(), "delegation", "add", image, "targets/releases", pubkey, "--all-paths"),
+		withNotaryPassphrase("foo"),
+	)
+}
+
+func TestPushWithContentTrustSignsAllFirstLevelRolesWeHaveKeysFor(t *testing.T) {}
+
+func TestPushWithContentTrustSignsForRolesWithKeysAndValidPaths(t *testing.T) {}
+
+func TestPullWithContentTrustSignsForRolesWithKeysAndValidPaths(t *testing.T) {}
```

Comment:
There are way too many e2e tests here. It must be possible for some of these to be unit tests.